sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -148,28 +482,50 @@ kmac_error_t kmac_update(const uint8_t *data, size_t data_len, uint8_t *digest,
     if (err != kKmacOk) {
       return err;
     }
-    abs_mmio_write8(
-        TOP_EARLGREY_KMAC_BASE_ADDR + KMAC_MSG_FIFO_REG_OFFSET + (i % 4),
-        *byte_ptr++);
+    abs_mmio_write8(kKmacBaseAddr + KMAC_MSG_FIFO_REG_OFFSET + (i % 4),
+                    *byte_ptr++);
+  }
+
+  // If operation=KMAC, then we need to write `right_encode(digest_len)`
+  if (operation == kKmacOperationKMAC) {
+    uint32_t digest_len_bit = 8 * digest_len;
+    if (digest_len_bit / 8 != digest_len) {
+      return kKmacDigestLenTooLongError;
+    }
+
+    uint8_t written_bytes = 0;
+    bool msbyte_found = false;
+    for (size_t i = 0; i < 4; ++i) {
+      uint8_t cur_byte = (digest_len_bit >> 8 * (3 - i)) & 0xff;
+      msbyte_found = msbyte_found || cur_byte || i == 3;
+      if (msbyte_found) {
+        abs_mmio_write8(kKmacBaseAddr + KMAC_MSG_FIFO_REG_OFFSET +
+                            ((data_len + written_bytes++) % 4),
+                        cur_byte);
+      }
+    }
```

Comment:
nit: I think it would be clearer to compute the length first as in `write_left_encode`. Maybe you could even have a more generic `little_endian_encode` function that returns the length and writes the value, and you could use it like:
```
// Left encode.
uint8_t len = little_endian_encode(buf + 1, value);
buf[0] = len;
memset(buf + len + 1, 0, ARRAYSIZE(buf) - (len + 1))

// Right encode.
uint8_t len = little_endian_encode(buf, value);
buf[len] = len;
memset(buf + len + 1, 0, ARRAYSIZE(buf) - (len + 1))
```

Reply:
This is a good suggestion, I will simplify both `right_encode` and `left_encode` with this. Thanks.